src/Microsoft.AspNetCore.Razor.Language/Intermediate/RazorIRNodeReference.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Razor.Language.Intermediate
+{
+    public struct RazorIRNodeReference
+    {
+        public RazorIRNodeReference(RazorIRNode parent, RazorIRNode node)
+        {
+            if (parent == null)
+            {
+                throw new ArgumentNullException(nameof(parent));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            Parent = parent;
+            Node = node;
+        }
+
+        public void Deconstruct(out RazorIRNode parent, out RazorIRNode node)
+        {
+            parent = Parent;
+            node = Node;
+        }
+
+        public RazorIRNode Node { get; }
+
+        public RazorIRNode Parent { get; }
```

Comment:
Looks like you don't even need to expose this guy. If you want to keep exposing it then all the methods on this could just be extension methods.